File: data_structures/hash_tables/hash_table.py
class HashTable:
    def __init__(self, size):
        self.size = size
        self.data = [None] * self.size
        
    def _hash(self, key):
        hash = 0
        for i in range(len(key)):
            hash += ord(key[i]) * i
        return hash % self.size

    def set(self, key, value):
        address = self._hash(key)
        if self.data[address] is None:
            self.data[address] = []
        self.data[address].append([key, value])
        return self.data

    def get(self, key):
        address = self._hash(key)
        bucket = self.data[address]
        if bucket:
            for k, v in bucket:
                if k == key:
                    return v
        return None
    
    def delete(self, key):
        address = self._hash(key)
        bucket = self.data[address]
        if bucket:
            for i in range(len(bucket)):
                if bucket[i][0] == key:
                    del bucket[i]
                    return True
        return False

    def keys(self):
        keys = []
        for i in range(len(self.data)):
            bucket = self.data[i]
            if bucket:
                for j in range(len(bucket)):
                    keys.append(bucket[j][0])
        return keys

File: data_structures/hash_tables/hash_table.js
class HashTable {
    constructor(size) {
        this.data = Array(size).fill(null);
    }

    _hash(key) {
        let hash = 0;
        for (let i =0; i < key.length; i++){
            hash = (hash + key.charCodeAt(i) * i) % this.data.length;
        }
        return hash;
    }

    set(key, value) {
        let address = this._hash(key);
        if (!this.data[address]) {
            this.data[address] = [];
        }
        this.data[address].push([key, value]);
        return this.data;
    }

    get(key) {
        let address = this._hash(key);
        let bucket = this.data[address];
        if (bucket) {
            for (let i=0; i < bucket.length; i++) {
                if (bucket[i][0] === key) {
                    return bucket[i][1];
                }
            }
        }
        return undefined;
    }

    delete(key) {
        let address = this._hash(key);
        let bucket = this.data[address];
        if (bucket) {
            for (let i=0; i < bucket.length; i++) {
                if (bucket[i][0] === key) {
                    let removedValue = bucket[i][1];
                    bucket.splice(i, 1);
                    return removedValue;
                }
            }
        }
        return undefined;
    }

    keys() {
        const keys = [];
        for (let i=0; i < this.data.length; i++) {
            let bucket = this.data[i];
            if(bucket){
                for (let j=0; j < bucket.length; j++) {
                    keys.push(bucket[j][0]);
                }
            }
        }
        return keys;
    }
}
